moveit_ros/planning/kinematics_plugin_loader/src/kinematics_plugin_loader.cpp

```diff
@@ -303,36 +346,100 @@ moveit::core::SolverAllocatorFn KinematicsPluginLoader::getLoaderFunction(const
+          const std::string ksolver_ik_links_param_suffix = ".kinematics_solver_ik_links";
+
+          // Base parameter name which prefixes other commands
+          // This may change depending on param search success
+          // NOTE: this is a messy pattern
+
+          const std::vector<std::string> base_param_names = {
+              // 1. Try to get param directly
+              known_group.name_,
+              // 2. Try to get param with alternate base name
+              robot_description_ + "_kinematics." + known_group.name_
+          };
+
+          for(const std::string &base_param_name : base_param_names)
+          {
+              if(ksolver_param.get_type() != rclcpp::ParameterType::PARAMETER_NOT_SET) {
+                  break;
+              }
+
+              std::string ksolver_param_name = base_param_name + ksolver_param_suffix;
+
+              RCLCPP_DEBUG(LOGGER, "Looking for param %s ", ksolver_param_name.c_str());
+
+              ksolver_param          = declare_parameter<rclcpp::ParameterType::PARAMETER_STRING>(node_, ksolver_param_name);
+              ksolver_timeout_param  = declare_parameter<rclcpp::ParameterType::PARAMETER_DOUBLE>(node_, base_param_name + ksolver_timeout_param_suffix);
+              ksolver_res_param      = declare_parameter<rclcpp::ParameterType::PARAMETER_DOUBLE>(node_, base_param_name + ksolver_res_param_suffix);
+              ksolver_ik_links_param = declare_parameter<rclcpp::ParameterType::PARAMETER_STRING_ARRAY>(node_, base_param_name + ksolver_ik_links_param_suffix);
+          }
+
+          // 3. Try to get it from the move_group node
           if (ksolver_param.get_type() == rclcpp::ParameterType::PARAMETER_NOT_SET)
           {
-            base_param_name = robot_description_ + "_kinematics." + known_group.name_;
-            ksolver_param_name = base_param_name + ".kinematics_solver";
-            RCLCPP_DEBUG(LOGGER, "Looking for param %s ", ksolver_param_name.c_str());
-            ksolver_param = declare_parameter<rclcpp::ParameterType::PARAMETER_STRING>(node_, ksolver_param_name);
+              std::string base_param_name = known_group.name_;
+
+              RCLCPP_DEBUG(LOGGER, "Looking for param %s in move_group node", (base_param_name + ksolver_param_suffix).c_str());
+
+              auto future = parameters_client->get_parameters({
+                      base_param_name + ksolver_param_suffix,
+                      base_param_name + ksolver_timeout_param_suffix,
+                      base_param_name + ksolver_res_param_suffix,
+                      base_param_name + ksolver_ik_links_param_suffix
+                      });
+
+              RCLCPP_DEBUG(LOGGER, "Waiting for param request future...");
+
+              std::future_status status;
+              do {
+                  status = future.wait_for(1s);
+              } while (status != std::future_status::ready);
+
+              ksolver_param          = future.get()[0];
```

Comment:
Better check if the future is valid fist. Also, I think it would be cleaner to check for vector size and use `at()`.